core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts

```diff
@@ -331,26 +308,65 @@ export class ConversationSkinComponent {
       })
     );
 
+    this.directiveSubscriptions.add(
+      this.conversationSkinService.onGiveFeedbackAndStayOnCurrentCard.subscribe(
+        args => {
+          this.giveFeedbackAndStayOnCurrentCard(
+            args.feedbackHtml,
+            args.missingPrerequisiteSkillId,
+            args.refreshInteraction,
+            args.refresherExplorationId
+          );
+        }
+      )
+    );
+
+    this.directiveSubscriptions.add(
+      this.conversationSkinService.onMoveToNewCard.subscribe(args => {
+        this.moveToNewCard(
+          args.feedbackHtml,
+          args.isFinalQuestion,
+          args.nextCard
+        );
+      })
+    );
+
+    this.directiveSubscriptions.add(
+      this.conversationSkinService.onShowUpcomingCard.subscribe(() => {
```

Comment:
This looks very strange. You say here "on showing an upcoming card this component should show the upcoming card". Huh?

Maybe you want to pass `this.showUpcomingCard` as a callback to the service, and the service should call it. But it does sound strange too, in my understanding the service should not be responsible for showing anything, it's not responsible for the view, it's responsible for the logic. 

The less callbacks you have in the module - the better. In this specific situation nothing stops you from returning something from handleFinalQuestion that tells the calling site if it should show the upcoming card or not. I suggest to not return boolean, as it says too little about what that boolean is, return an enum with two values.

You worked with this part more, maybe my suggestion doesn't work or is difficult to implement, if so, just tell me. My main points are:
1. These names are strange, they don't make much sense.
2. The less callbacks you have - the better. Try to pass that information through the return value.